src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs

```diff
@@ -84,6 +91,100 @@ public HttpClient CreateClient()
                 handler.ServerCertificateCustomValidationCallback = (req, cert, chain, errors) => true;
 #endif
             }
+            // If schannel is the TLS backend, custom certificate usage must be explicitly enabled
+            else if (!string.IsNullOrWhiteSpace(_settings.CustomCertificateBundlePath) &&
+                ((_settings.TlsBackend != TlsBackend.Schannel) || _settings.UseCustomCertificateBundleWithSchannel))
+            {
+                string certBundlePath = _settings.CustomCertificateBundlePath;
+                _trace.WriteLine($"Custom certificate verification has been enabled with certificate bundle at {certBundlePath}");
+
+                // Throw exception if cert bundle file not found
+                if (!_fileSystem.FileExists(certBundlePath))
+                {
+                    throw new FileNotFoundException($"Custom certificate bundle not found at path: {certBundlePath}", certBundlePath);
+                }
+
+                Func<X509Certificate2, X509Chain, SslPolicyErrors, bool> validationCallback = (cert, chain, errors) =>
+                {
+                    // Fail immediately if there are non-chain issues with the remote cert
+                    if ((errors & ~SslPolicyErrors.RemoteCertificateChainErrors) != 0)
+                    {
+                        return false;
+                    }
+
+                    // Import the custom certs
+                    X509Certificate2Collection certBundle = new X509Certificate2Collection();
+                    certBundle.Import(certBundlePath);
+
+                    try
+                    {
+                        // Add the certs to the chain
+                        chain.ChainPolicy.ExtraStore.AddRange(certBundle);
+
+                        // Rebuild the chain
+                        if (chain.Build(cert))
+                        {
+                            return true;
+                        }
+                        else if (chain.ChainStatus.Any(status => status.Status != X509ChainStatusFlags.UntrustedRoot))
+                        {
+                            // Fail for any errors other than untrusted root
+                            return false;
+                        }
+                        else if (chain.ChainStatus.Any(status => status.Status == X509ChainStatusFlags.UntrustedRoot))
+                        {
+                            // Manually verify root is contained within the certBundle
+                            X509Certificate2 rootCert = chain.ChainElements[chain.ChainElements.Count - 1].Certificate;
```

Comment:
Is it ever possible to have a chain with zero `ChainElements`? (such that `.Count - 1 < 0`)

Reply:
Apparently, [it is possible](https://social.msdn.microsoft.com/Forums/en-US/d7f72468-d8fb-4a37-8be2-82c5ca62485c/httpclienthandlerservercertificatecustomvalidationcallback-receives-empty-certchain?forum=xamarinios) (although, to me, what's described there sounds more like a bug than an intentional feature, and I only saw a reference to it for iOS). However, because this block is predicated on having an Untrusted Root error thrown by the previous chain build, I think it would need at least one cert (the root) in the chain.

Reply:
Sounds good to me!  Should be fine to leave as is.